feat(ui): inline virtual text

bfredl opened the pull request:
Note: This is just the first patch. there are many vim patches to go through which fixes various edge cases and bugs.

zeertzjq commented on src/nvim/plines.c:
```diff
@@ -296,8 +296,16 @@ void init_chartabsize_arg(chartabsize_T *cts, win_T *wp, linenr_T lnum FUNC_ATTR
   cts->cts_line = line;
   cts->cts_ptr = ptr;
   cts->cts_cur_text_width = 0;
-  // TODO(bfredl): actually lookup inline virtual text here
```
The doxygen comment above this function may also need change to stop mentioning "text properties", and `FUNC_ATTR_UNUSED` can be removed now.

Jesse-Bakker commented:
On difference between this and the included vim patches seems to be redrawing the cursor after inserting inline virtual text before it.

A failing test case:
```lua
    it('adjusts cursor location correctly when inserting around inline virtual text', function()
        insert('12345678')
        feed '$'
        meths.buf_set_extmark(0, ns, 0, 4,
            { virt_text = { { ' virtual text ', 'Special' } }, virt_text_pos = 'inline' })

        screen:expect { grid = [[
      1234{28: virtual text }567^8                            |
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
      {1:~                                                 }|
                                                        |
      ]]
        }
end)
```

Instead of drawing the cursor at the end of the line, it is drawn in the middle of the virtual text until it is updated by moving the cursor left or entering insert mode.

xiantang commented:
need to consider the case of `gj` and `gk` which are moved between the display line, I patched this pr, but actually, it works strangely. I supposed need to consider the virtual text in display line.

Jesse-Bakker commented:
A dirty fix:

```diff
diff --git a/src/nvim/decoration.c b/src/nvim/decoration.c
index 037eb9f0d..96895bc1b 100644
--- a/src/nvim/decoration.c
+++ b/src/nvim/decoration.c
@@ -78,6 +78,13 @@ void decor_redraw(buf_T *buf, int row1, int row2, Decoration *decor)
       redraw_buf_range_later(buf, row1 + 1, row2 + 1);
     }
   }
+  if (decor && decor_virt_pos(*decor) && decor->virt_text_pos == kVTInline) {
+        FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
+            if (wp->w_cursor.lnum == row1 + 1) {
+                wp->w_valid &= ~(VALID_VIRTCOL | VALID_WROW | VALID_WCOL);
+            }
+        }
+    }

   if (decor && decor_virt_pos(*decor)) {
     redraw_buf_line_later(buf, row1 + 1, false);
```

rcarriga commented:
Lots of discussion here so bit of an info dump, apologies.

> We could even just return the cancel function directly but returning a table leaves it more open to extending in future.

Not sure what I was thinking here when I said this. We can just return a task ID and provide module level functions for the cancelling etc, which would probably fit much more with the rest of core's APIs.

```lua
local task_id = async.run(function() end)
async.cancel(task_id)
```

---


> Hmm, I would try and tweak this to something like:

I'm not sure overloading a `pcall` function would make the most sense. There's a a couple of points that stick out to me:
Passing a task to pcall is a bit deceptive because the task is already running. It would also add a fair bit of boilerplate if making several calls and when adding arguments

```lua
local success, result1 = async.pcall(async.run(async_func_1))

local success, result2 = async.pcall(async.run(function()
  async_func_2(arg1)
end))

local function proxying_func(...)
  local args = {...}
  local success, result = async.pcall(async.run(function()
    async_func_3(unpack(args))
  end))
end
```

If we just pass async functions directly, which would be common in async contexts then we have to decide how to handle arguments

We could allow passing arguments
```lua
local success, result1 = async.pcall(async_func_1)

local success, result2 = async.pcall(async_func_2, arg1)

local function proxying_func(...)
  local success, result = async.pcall(async_func_3, ...)
end
```

Which is great but if we want to allow calling it in a non async context then we'd either have to require the number of arguments to be provided or provide the callback before the arguments

```lua
-- Provide number of arguments
async.pcall(async_func_1, 0, function(success, result))
    ...
end)

async.pcall(async_func_2, 1, arg1, function(success ,result)
    ...
end)

-- Provide callback before args
async.pcall(async_func_1, function(success, result))
    ...
end)

async.pcall(async_func_2, function(success ,result)
    ...
end, arg1)
```

Both of which require changing the type signature in IMO a clunky way that would not play well with the LSP. Also might lead to some surprising behaviour when calling from an async context but trying to provide a callback.

Sorry to be so verbose on this, just wanted to illustrate my thoughts. I think keeping `pcall` as an async only function makes the API much simpler.

---

> The resolution was that we should just be more strict on naming conventions and/or annotate the code better with things like -- @async

Will also add that the Lua LSP has an inline hint for `await` when calling functions marked as `async` so users can use that to see awaits easily. When https://github.com/neovim/neovim/pull/20130 is merged, it'll be an even better experience as they can be inline.

SleepySwords committed:
feat(ui): inline virtual text

vim-patch:9.0.0067: cannot show virtual text

Problem:    Cannot show virtual text.
Solution:   Initial changes for virtual text support, using text properties.

https://github.com/vim/vim/commit/7f9969c559b51446632ac7e8f76cde07e7d0078d

vim-patch:9.0.0116: virtual text not displayed if 'signcolumn' is "yes"

Problem:    Virtual text not displayed if 'signcolumn' is "yes".
Solution:   Set c_extra and c_final to NUL.
https://github.com/vim/vim/commit/711483cd1381a4ed848d783ae0a6792d5b04447b

Co-authored-by: bfredl <bjorn.linse@gmail.com>

SleepySwords committed:
fix(ui): fix cursor position with multiple inline virtual text

vim-patch9.0.0121: cannot put virtual text after or below a line

Problem:    Cannot put virtual text after or below a line.
Solution:   Add "text_align" and "text_wrap" arguments.

https://github.com/vim/vim/commit/b7963df98f9dbbb824713acad2f47c9989fcf8f3

This only patches the fix, not the whole thing.

SleepySwords committed:
vim-patch:9.0.0130: cursor position wrong when inserting around virtual text

Problem:    Cursor position wrong when inserting around virtual text.
Solution:   Update the cursor position properly.

https://github.com/vim/vim/commit/1f4ee19eefecd8f70b7cbe8ee9db8ace6352e23e

Co-authored-by: tom-anders <13141438+tom-anders@users.noreply.github.com>

SleepySwords committed:
fix(ui): adjust the cursor when inserting virtual text

Credit to: Jesse Bakker

https://github.com/neovim/neovim/pull/20130#issuecomment-1369652743

Co-authored-by: Jesse Bakker <github@jessebakker.com>

SleepySwords committed:
fix(ui): fix multi-byte characters highlight in virtual text

This also fixes insert cursor position around virtual text

vim-patch:9.0.0132: multi-byte characters in virtual text not handled correctly

Problem:    Multi-byte characters in virtual text not handled correctly.
Solution:   Count screen cells instead of bytes.

https://github.com/vim/vim/commit/09ff4b54fb86a64390ba9c609853c6410ea6197c

SleepySwords committed:
vim-patch:9.0.0143: cursor positioned after virtual text in empty line

Problem:    Cursor positioned after virtual text in empty line.
Solution:   Keep cursor in the first column. (closes vim/vim#10786)

https://github.com/vim/vim/commit/afd2aa79eda3fe69f2e7c87d0b9b4bca874f386a

SleepySwords committed:
vim-patch:9.0.0179: cursor pos wrong with wrapping virtual text in empty line

Problem:    Cursor position wrong with wrapping virtual text in empty line.
Solution:   Adjust handling of an empty line. (closes vim/vim#10875)

https://github.com/vim/vim/commit/49a90792d950c51608d0459ef8699fe921070718

Co-authored-by: Bram Moolenaar <Bram@vim.org>

SleepySwords committed:
vim-patch:9.0.0178: cursor position wrong with virtual text before Tab

Problem:    Cursor position wrong with virtual text before Tab.
Solution:   Use the byte length, not the cell with, to compare the column.
            Correct tab size after text prop. (closes vim/vim#10866)

https://github.com/vim/vim/commit/e428fa04a758cc87ea580c856a796e58e407504b

Co-authored-by: Bram Moolenaar <Bram@vim.org>

SleepySwords committed:
vim-patch:9.0.0183: extra space after virtual text when 'linebreak' is set

Problem:    Extra space after virtual text when 'linebreak' is set.
Solution:   Do not count virtual text when getting linebreak value.
            (closes vim/vim#10884)

https://github.com/vim/vim/commit/52de3a8d3943520bbd4e5e40a4c43fcc7182dac0

Co-authored-by: Bram Moolenaar <Bram@vim.org>

SleepySwords committed:
fix(ui): fix visual and search highlighting interfering with virtual text

vim-patch:9.0.0193: search and match highlgith interfere with virtual text

Problem:    Search and match highlgith interfere with virtual text highlight.
            (Ben Jackson)
Solution:   Check for match highlight after text properties.  Reset and
            restore search highlight when showing virtual text.
            (closes vim/vim#10892)

https://github.com/vim/vim/commit/e38fc86180fd3f6b372648eea6adc3f623fea302

vim-patch:9.0.0452: Visual highlighting extends into virtual text prop

Problem:    Visual highlighting extends into virtual text prop.
Solution:   Do not highlight what isn't actually selected.  Fix ordering of
            stored text props.

https://github.com/vim/vim/commit/6eda17d881c9b2880ccb2a4d11951939a58f233d

Co-authored-by: Bram Moolenaar <Bram@vim.org>

SleepySwords committed:
vim-patch:9.0.0205: cursor in wrong position when inserting after virtual text

Problem:    Cursor in wrong position when inserting after virtual text. (Ben
            Jackson)
Solution:   Put the cursor after the virtual text, where the text will be
            inserted. (closes vim/vim#10914)

https://github.com/vim/vim/commit/28c9f895716cfa8f1220bc41b72a534c0e10cabe

Co-authored-by: Bram Moolenaar <Bram@vim.org>

SleepySwords committed:
vim-patch:9.0.0210: 'list' mode does not work properly with virtual text

Problem:    'list' mode does not work properly with virtual text.
Solution:   Show the "$" at the right position. (closes vim/vim#10913)

https://github.com/vim/vim/commit/c3a483fc3c65f649f9985bb88792a465ea18b0a2#diff-15009492c2b1d0a2629908b4618ad51c99d16746f238a0e6451dfe32355ed32fR1653

This commit only contains the seemingly refactored portion (not entirely
sure what it does)

Co-authored-by: Bram Moolenaar <Bram@vim.org>

SleepySwords committed:
fix(ui): fixes incorrect rendering when virtual text is not visable and nowrap

SleepySwords committed:
vim-patch:9.0.0716: with 'nowrap' virtual text "after" does not scroll left

Problem:    With 'nowrap' virtual text "after" does not scroll left.
Solution:   Skip part of the virtual text that is left of the window.
            (closes vim/vim#11320)  Fix going beyond the last column of the window.

https://github.com/vim/vim/commit/cd105417a53fcf97c0935f3468201ef11516c9f1

Co-authored-by: Bram Moolenaar <Bram@vim.org>

SleepySwords committed:
vim-patch:9.0.0944: 'cursorline' causes virtual text highlight to continue

Problem:    'cursorline' causes virtual text highlight to continue.
Solution:   Save and restore line_attr. (closes vim/vim#11588)

https://github.com/vim/vim/commit/6ac16f0c0fe923098b9df5ac430f1923045f16ea

Co-authored-by: Bram Moolenaar <Bram@vim.org>

SleepySwords committed:
vim-patch:9.0.1067: in diff mode virtual text is highlighted incorrectly

Problem:    In diff mode virtual text is highlighted incorrectly. (Rick Howe)
Solution:   Do not use diff attributes for virtual text. (closes vim/vim#11714)

https://github.com/vim/vim/commit/d097af77797f030e0f29f9bbc298358a5addb2a5

Co-authored-by: Bram Moolenaar <Bram@vim.org>

SleepySwords committed:
fix(ui): fix incorrect highlighting when virtual text wraps with number

SleepySwords committed:
fix(ui): fix incorrect highlighting when virtual text next to match

SleepySwords committed:
fix(ui): fix tabs not being spaced properly after virtual text with no wrap

also fixes incorrect skipping of multibyte characters

SleepySwords committed:
fix(ui): fix virtual text not displaying when two overlapping inlines (nowrap)

SleepySwords committed:
fix(ui): fix overflowing nowrap virtual text not displaying if tab follows

tom-anders commented:
@bfredl I'd like to help with this PR if possible, is there a list of vim patches that we need to port over for this? Probably just `git log --grep="virtual text"` should be a good start?

glepnir commented:
as far as i know in vim it's `textprop`

luukvbaal commented:
Grepping the git log is a start but not every commit message for a relevant patch is guaranteed to contain those patterns. I think there is no other way but to actually skim through the entire log and make a judgement call of whether it could be relevant before inspecting the commit.

At least that's what I did for porting over the 'smoothscroll' [patches](https://github.com/neovim/neovim/pull/23320/commits), there are several that don't contain the word 'smoothscroll'.

tom-anders commented:
I think I'll start with trying to fix https://github.com/neovim/neovim/issues/20524 - There's a corresponding functional test case in this PR that currently fails (`fold attached virtual lines are drawn correctly #21837`)

SleepySwords commented:
If you want to take a look, I have a [branch](https://github.com/SleepySwords/neovim/tree/inline2) that I believe contains some of the main patches (I probably did miss quite a bit, this was only searching `virtual text` on GitHub commits, there are also some patches that I still am not sure if they need to be implemented). It is functional enough that I sometimes use this for inlay hints while working on projects. Not too sure if the code quality is up to standard. There are also still a lot of edge cases that need to be fixed.

Some stuff that is still broken/implemented:
- [ ] RHS does not seem to work
- [ ] If a virtual text is slightly, but not fully overlapping with `showbreak` enabled, the alignment of the first character is off. (happens in vim)
- [ ] Emphermal does not work (Not sure what the solution might be?)
- [ ] If `list` is enabled with an extend char and the virtual text is perfectly outside, the extend character is not displayed (also happens in vim)
- [ ] Deleting characters causes weird drawing (https://github.com/vim/vim/issues/12244)
- [ ] Probably some other extmark options (is something like `hl_mode` relevant?)

Probably even more that are not found yet, especially with multiple virtual texts.
Just as a note: while fixing failing tests, I messed up the initial commit, which is then fixed later (https://github.com/neovim/neovim/commit/98ac518f74f6d8352a266acd448672b57ae1a9b2)

Hope this helps!

bfredl commented:
@SleepySwords great work! 

Not everything needs to work in a MVP PR. We should start with completely disallowing inline virtual text in ephemeral marks (and then consider limited functionality as a follow up).

Not all edge cases need to be "right" immediately. What's important is than we do have tests for the ones that work (by having specific code in patches, etc). Then we can get something merged and iterate over the 0.10 cycle.

tom-anders commented:
> If you want to take a look, I have a [branch](https://github.com/SleepySwords/neovim/tree/inline2) that hopefully contains most of the main patches (I probably did miss quite a bit, this was only searching `virtual text` on GitHub commits, where some patches were skipped since I still was not sure if they need to be implemented or were relevant). It is functional enough that I sometimes use this for inlay hints while working on projects. Not too sure if the code quality is up to standard, and probably does contain dodgy fixes. There are also still a lot of edge cases that need to be fixed (RHS doesn't work, ephemeral doesn't work, weird edge cases regarding `showbreak` and `list with extends`, etc).
> 
> Just as a note: while fixing failing tests, I messed up the initial commit, which is then fixed later [98ac518](https://github.com/neovim/neovim/commit/98ac518f74f6d8352a266acd448672b57ae1a9b2) (Kinda dodgy fix)
> 
> Hope this helps!

Thanks! Just for my understanding: what does RHS refer to in this context?

SleepySwords commented:
> Thanks! Just for my understanding: what does RHS refer to in this context?

Sorry I meant to say `rightleft`. I think I have misunderstood how `rightleft` is supposed to work and it seems fine? Is inserting one character after the cursor expected behaviour? It looked a bit odd with the vertical bar cursor.

tom-anders committed:
docs: update api.txt and add inline virtual text to news.txt

bfredl committed:
fix(extmark): fix cursor position with both left and right gravity inline text

bfredl committed:
fix(test): clean up inline virtual text tests a little

zeertzjq commented on src/nvim/plines.h:
```diff
@@ -11,9 +11,12 @@ typedef struct {
+  int cts_cur_text_width_left;     // width of virtual text left of cursor
+  int cts_cur_text_width_right;     // width of virtual text right of cursor
+  MarkTreeIter cts_iter[1];
   // TODO(bfredl): iterator in to the marktree for scanning virt text
```
```suggestion
```

bfredl committed:
refactor(drawline): move inline text code to its own function

bfredl commented:
Merged, thanks @SleepySwords , @tom-anders and everyone else who have contributed with code and testing!

As always with a master feature, API and feature details is subject to change until next release (0.10) feature freeze. With this in mind, early testing and bug reports from early plugin adopters would be very welcome.

nyngwang commented:
Oh my god, thank you so much!!!!

B4mbus commented:
Finally! Now I can die in peace!

clason commented:
So can this PR.
